download_plugins.py

```diff
@@ -1,84 +1,124 @@
+import requests
+import yaml
+import sys
+from textwrap import dedent
+from distutils.version import LooseVersion
+
 
 # Get dependencies for a plugin, or get depedencies for a dependency.
 def get_dependencies(plugin):
-    print "Download dependencies for:", plugin
-    print "*****************************************"
-    print plugin, "has these dependencies:"
+    print(dedent("""
+    Finding dependencies for: {0}
+    ******************************************
+    {0} has these dependencies:""".format(plugin)))
 
     try:
-        dependencies = plugins_list['plugins'][plugin]['dependencies']
+        dependencies = plugins_list["plugins"][plugin]["dependencies"]
     except KeyError:
-        print "Unable to find dependencies for %s." % plugin
+        print("Unable to find dependencies for {}.".format(plugin))
+        global exit_code
         exit_code = 1
         return None
-    
+
     for dependency in dependencies:
-        print "Processing dependency: " + dependency['name']
-        
-        if (dependency['name'] not in installed_plugins) and (not dependency['optional']):
-            print "installing " + dependency['name']
+        print("Processing dependency: {}".format(dependency["name"]))
+
+        if dependency["optional"]:
+                print("{} is optional".format(dependency["name"]))
+                continue
+
+        elif dependency["name"] in stored_plugins:
+            if (LooseVersion(dependency["version"]) <=
+                    LooseVersion(stored_plugins[dependency["name"]])):
+                print("Newer or same version of dependency already added")
+            else:
+                print("Adding new version of dependency")
+                stored_plugins[dependency["name"]] = dependency["version"]
+                get_dependencies(dependency["name"])
+            if dependency["name"] not in multi_version_plugins:
+                    multi_version_plugins.append(dependency["name"])
 
-            download_plugin(dependency['name'])
-            get_dependencies(dependency['name'])
         else:
-            if dependency['optional']:
-                print dependency['name'] + " is optional"
-            else: 
-                print dependency['name'] + " already downloaded"
+            print("Adding dependency: {}".format(dependency["name"]))
+            stored_plugins[dependency["name"]] = dependency["version"]
+            get_dependencies(dependency["name"])
 
-# Download plugin from jenkins update server. If no version is specified, latest plugin is downloaded.
-def download_plugin(plugin, version = None):
+
+# Download plugin from jenkins update server.
+# If no version is specified, latest plugin is downloaded.
+def download_plugin(plugin, version=None):
     download_url = ""
-    
+
     if version is None:
-        download_url = plugin_base_url + "/latest/%s.hpi" % plugin
+        download_url = (plugin_base_url +
+                        "/latest/{}.hpi".format(plugin))
```

Comment:
The line is not long enough to break to a new line. Also, you could probably just only use `format` for structure the URL here.